Implement the o-autocomplete design using accessible-autocomplete

JakeChampion opened the pull request:
[The design in Figma.](https://www.figma.com/file/RCQDZBkQSCtzs5DbUWCldu/organisation-name)

The logic behind the autocomplete is handled mostly by [accessible-autocomplete](https://github.com/alphagov/accessible-autocomplete). There is a [bug in accessible-autocomplete](https://github.com/alphagov/accessible-autocomplete/pull/491) which meant we needed to fork the project, the project doesn't receive updates right now, which means we are not missing out on other changes to the project by forking it.

Screenshot of a single suggestion being displayed:
<img width="248" alt="image" src="https://user-images.githubusercontent.com/1569131/122046122-6bc99300-cdd6-11eb-9a41-9a92dd666530.png">

Screenshot of multiple suggestions being displayed:
<img width="248" alt="image" src="https://user-images.githubusercontent.com/1569131/122046156-771cbe80-cdd6-11eb-8ba7-c9e2cc7601f5.png">


The tests are written using [Testing Library](https://testing-library.com/docs/), which helps write tests in a way that is close to what a user would do, such as tabbing, typing into inputs, press enter on an element in focus, etc.

JakeChampion committed:
Add autocomplete npm package as a dependency

JakeChampion committed:
Add initial attempt of wrapping Origami around gds autocomplete component

JakeChampion committed:
stash

JakeChampion committed:
update to prereleases of the components

JakeChampion committed:
add branding for search colors

JakeChampion committed:
start using the mixins from origami instead of duplicating their styles via copy/paste

JakeChampion committed:
add missing visual-effects dependency

JakeChampion committed:
remove single space indent to make `obt verify` happy

JakeChampion committed:
use o-forms in the demos to provide the label styling

JakeChampion committed:
remove duplicated styles and remove styles which come from o-forms-input--text

JakeChampion committed:
do not allow autoselect to be turned on, until we edcide whether to support it

JakeChampion committed:
add the notes file

notlee committed:
wip: styles & notes

JakeChampion committed:
update the notes

JakeChampion committed:
Make the clear button be styled correctly for the enhanced select element

enhanced select elements have a wrapper div added to the dom by accessible-autocomplete, this causes issues with our current css implementation for the clear button, which did not work with an extra wrapper div.
I've updated the css to expect an extra wrapper div and updated the enhanced text input mode to add the same wrapper div that the enhaced select mode has, this way they have the same dom structure after accessible-autocomplete is executed.

JakeChampion committed:
remove the source data attribute as enhanced select mode does not work with this attribute

JakeChampion committed:
change whitespace from spaces to tabs to adhere to the components linting rules

JakeChampion committed:
remove brand variables which are not used

JakeChampion committed:
update the package and origami manifests to follow the spec 2 updates

JakeChampion committed:
use our own container to be protected from accessible-autocomplete chaning their dom structure

instead of relying on the dom structure that accessible-autocomplete creates, we now provide our our container div, which provides the grid for the autocomplete and clear button to align within

JakeChampion committed:
add description to the static data markup example

JakeChampion committed:
Add delat to custom suggestions demo to help with working on the loading design

JakeChampion committed:
fix highlighting bug where characters which didn't match would get highlighted

JakeChampion committed:
make the results menu padding match the design in figma

JakeChampion committed:
add missing semicolon

JakeChampion committed:
Add a loading state when waiting for results to populate the results menu

JakeChampion committed:
Add a demo which delays populating the results menu to test the loading state

JakeChampion committed:
Make all the demos use the same suggestions and suggestion filter logic

JakeChampion committed:
Add keyboard support to the readme as I keep forgetting what each key is meant to do

JakeChampion committed:
clarify that pressing Enter only does nothing if the results menu is open and the focus is within the text input

JakeChampion committed:
Add to the readme that we built this on top of accessible autocomplete

JakeChampion committed:
update documentation for the tab key to correctly document that it moves focus to the clear button

JakeChampion committed:
Add documentation for pressing backspace key when focus is within the results menu

JakeChampion committed:
Add keyboard documentation for the clear button

JakeChampion committed:
Add documentation for the dynamic suggestions use-case

JakeChampion committed:
Add example code for custom/dynamic suggestions

JakeChampion committed:
rename customSource to customSuggestions to match documentation

JakeChampion committed:
correct the sentence for the markup documentation on dynamic suggestions

JakeChampion committed:
add jsdoc for customSuggestions

JakeChampion committed:
rename demos

JakeChampion committed:
Hide the suggestions when loading more suggestions

do the hiding using our own classname and not one which is controlled by the accessible-autocomplete preact element

JakeChampion committed:
move example code aboce the documentation

JakeChampion committed:
use the debounce function from o-utils

JakeChampion committed:
update the table of contents

JakeChampion committed:
add the missing fields which obt v reported

JakeChampion commented:
https://user-images.githubusercontent.com/1569131/119343071-daab4480-bc8d-11eb-9829-ec57e413cc90.mov

JakeChampion committed:
only run the most recent suggestion api call

JakeChampion committed:
update assertions nd fixtures for basic tests

JakeChampion committed:
Delete notes.txt

JakeChampion committed:
remove redundant `for` attribute on labels

the labels are wrapping their inputs, which means they do not need a `for` attribute

JakeChampion committed:
Add tests for enhanced select element mode

JakeChampion committed:
Add link to PR which would remove the need for the hack file -- https://github.com/facebook/jest/pull/11569

JakeChampion committed:
add keyboard navigation and clear button tests for enhanced select mode

JakeChampion commented on test/scss/_main.test.scss:
```diff
@@ -7,9 +7,10 @@
-				@include expect($selector: false) {
-					.test-mixin .o-autocomplete {
-						display: block;
+				@include contains($selector: false) {
```
Question: Does this component need sass tests?

notlee replied:
I would say not, since there is only one primary mixin with no options -- Percy can catch visual changes

JakeChampion commented on src/js/autocomplete.js:
```diff
@@ -6,9 +13,187 @@ class Autocomplete {
+		// If highlighter is a string, then it is the name of a global function to use.
+		// If highlighter is not a string, then it is a function to use.
+		if (typeof this.options.highlighter === 'string') {
+			this.options.highlighter = window[this.options.highlighter];
```
TODO: Need to write a test for the highlighter function

JakeChampion commented on src/js/autocomplete.js:
```diff
@@ -6,9 +13,187 @@ class Autocomplete {
+			 * @param {PopulateResults} populateResults - Function to call when ready to update the suggestions dropdown
+			 * @returns {void}
+			 */
+			const customSuggestions = typeof this.options.source === 'string' ? window[this.options.source] : this.options.source;
```
TODO: Need to write a test for the source function

JakeChampion replied:
Done in [`2dfd0d2` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/2dfd0d273e3de893f4f45931b44f1a704cdc7fa4)

notlee commented on README.md:
```diff
@@ -1,11 +1,18 @@
 # o-autocomplete
 
-An origami component for autocomplete inputs
+An origami component for autocomplete inputs. This is built on top of the excellent [accessible autocomplete component](https://github.com/alphagov/accessible-autocomplete) by AlphaGov.
```
```suggestion
An Origami component for autocomplete inputs. This is built on top of the excellent [accessible autocomplete component](https://github.com/alphagov/accessible-autocomplete) by AlphaGov.
```

notlee commented on README.md:
```diff
@@ -16,11 +23,28 @@ Check out [how to include Origami components in your project](https://origami.ft
 
+### For a static set of suggestions
+
+To provide a static set of suggestions, it is recommended to provide the suggestions via a `select` element, o-autocomplete will then progressively enhance the `select` element, using the provided `option` elements as the source for the suggestions.
```
"it is recommended" is an example of passive voice, which we try to avoid:
https://origami.ft.com/docs/principles/tone-and-language/#use-the-active-voice

```suggestion
To provide a static set of suggestions, we recommend using a `select` element. o-autocomplete will progressively enhance the `select` element, using the provided `option` elements as the source for the suggestions.
```

notlee commented on README.md:
````diff
@@ -16,11 +23,28 @@ Check out [how to include Origami components in your project](https://origami.ft
+```html
+<div data-o-component="o-autocomplete" class="o-autocomplete" id="my-autocomplete"></div>
+```
+
````
We need to update the Sass example to import the `main` file now
```scss
@import "@financial-times/o-autocomplete/main";
```
(sorry I can't make a suggestion as it's not changed in this PR)

JakeChampion replied:
Done in [`d12f692` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/d12f6925f976ad6c4e1d21391116a1b6a4403b0a)

notlee commented on README.md:
```diff
@@ -52,6 +76,67 @@ oAutocomplete.init(oAutocompleteElement);
 
 [Learn more about Origami component initialisation](https://origami.ft.com/docs/components/initialising/).
```
The initialisation page [is being replaced](https://github.com/Financial-Times/origami-website/pull/360/commits/d447ba3f153a4afa33d868014064f24e05310a20), since the init method is no longer covered by a spec and was optional. Maybe a link to the component's JSDoc would be better?

notlee commented on README.md:
````diff
@@ -52,6 +76,67 @@ oAutocomplete.init(oAutocompleteElement);
+#### Example
+
+```js
+async function customSuggestions(query, populateResults) {
````
It's not clear what to do with the custom suggestions function. A full example that includes passing the function to o-autocomplete would be excellent

notlee replied:
I find the parameter/return table makes the readme harder to browse and maybe a little intimidating, there's a lot to parse on one page. I prefer to focus on examples by usecase in the README and leave the full API details to JSDoc. That might be more of a personal preference though!

JakeChampion replied:
I've updated the example to show how to setup o-autocomplete with the customSuggestions function.

I'm not sure about the parameter/return tables either, I'm happy to remove them and lean on the JSDoc 👍

notlee commented on README.md:
```diff
@@ -52,6 +76,67 @@ oAutocomplete.init(oAutocompleteElement);
+| results | <code>Array.&lt;string&gt;</code> | The results to show in the suggestions menu |
+
+
+## Keyboard Support
```
Helpful!

notlee commented on demos/src/dynamic-delayed/dynamic-delayed.mustache:
```diff
@@ -0,0 +1,14 @@
+        </span>
+        <span class="o-forms-input o-forms-input--text">
+            <div data-o-component="o-autocomplete" data-o-autocomplete-source="customSuggestions" data-o-autocomplete-highlighter="customHighlighter" class="o-autocomplete" id="my-autocomplete">
+                {{! If the JavaScript executes, then this input will be progressively enhanced to an autocomplete component }}
```
I think this comment will be compiled away, and should be a HTML comment instead to show in the registry demo html?

JakeChampion replied:
```suggestion
                <!-- If the JavaScript executes, then this input will be progressively enhanced to an autocomplete component -->
```

notlee commented on README.md:
````diff
@@ -16,11 +23,28 @@ Check out [how to include Origami components in your project](https://origami.ft
+    </select>
 </div>
 ```
 
````
How do we document the o-forms case, as in the demo? Maybe a separate markup example with o-forms markup and a paragraph saying form CSS needs to be included?

JakeChampion replied:
To confirm I understand what you are saying, do you want the readme to include an example of how to use o-autocomplete as an input within o-forms?

JakeChampion replied:
Done in [`95d084b` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/95d084bafc23611ef94cd394d3432e9fe63ebed7)

notlee commented on main.scss:
```diff
@@ -12,14 +20,154 @@
 @mixin oAutocomplete ($opts: ()) {
 	// content of primary mixin
-	.o-autocomplete {
+	.o-autocomplete .o-autocomplete__container {
+		display: inline-grid;
+		position: relative;
+	}
+
+	.o-autocomplete .o-autocomplete__container > div:first-child {
+		@include oTypographySans($scale: 0);
+		grid-row: 1;
+		grid-column: 1;
+	}
```
I think there's an opportunity to reduce some specificity here, by not applying layout styles to alphagov elements. 

```suggestion
	.o-autocomplete {
		display: inline-grid;
		position: relative;
	}

	.o-autocomplete__container {
		@include oTypographySans($scale: 0);
		grid-row: 1;
		grid-column: 1;
	}
```

For the above to work the `addOrRemoveClearButton` function would also need to be updated to append the clear button to the root o-autocomplete element rather than the alphagov container. 

With those changes the close button will be miss-aligned on the loading state. To fix that we'd need to make the loading state absolutely positioned in the same way as the result list. Which will also avoid pushing page content down when loading (see how the lorem text is shifted down on loading in the example below).

https://user-images.githubusercontent.com/10405691/122070487-eb169100-cded-11eb-876b-e9942e84c786.mp4

notlee replied:
Another option would be to do the above but remove the container, styling `.o-autocomplete__wrapper` instead. Then the container/wrapper nonsense naming is avoided but there's a greater chance alphagov changes will conflict with o-autocomplete in the future

JakeChampion replied:
> Another option would be to do the above but remove the container, styling `.o-autocomplete__wrapper` instead. Then the container/wrapper nonsense naming is avoided but there's a greater chance alphagov changes will conflict with o-autocomplete in the future

We can avoid the potential conflict issue by pinning to a specific version of accessibile-autocomplete and updating to newer versions manually. That way Percy would catch any conflicts for us.

Which approach would you recommend going with?

notlee replied:
👍 let's remove the container and pin

notlee replied:
As requested, I had a go at this :) https://github.com/Financial-Times/o-autocomplete/pull/3

notlee commented on main.scss:
```diff
@@ -12,14 +20,154 @@
+	.o-autocomplete .o-autocomplete__menu-loading {
+		grid-row: 2;
+	}
+	.o-autocomplete .o-autocomplete__menu-loading-container {
```
This should be absolutely positioned to avoid reflow of page content (see above)

notlee commented on src/js/autocomplete.js:
```diff
@@ -6,9 +13,187 @@ class Autocomplete {
+			autocompleteEl.innerHTML = '';
+			autocompleteEl.appendChild(container);
+			const id = autocompleteEl.getAttribute('id');
+			autocompleteEl.removeAttribute('id');
```
Maybe we should generate a new `id` here? I think as a user I would be confused if an `id` I set in markup moved.

JakeChampion replied:
The reason we use the same `id` is because it could be linked to from a `label` element's `for` attribute. If we generate a new `id`, I think we would then need to find if a label exists which has a `for` attribute with the old `id` and update the attribute to use the new `id`?

notlee replied:
Oh I see, good point! In that case the the `id` should be set on the fallback input instead, so it's connected to the label if js fails.

```diff
-            <div data-o-component="o-autocomplete" data-o-autocomplete-source="customSuggestions" data-o-autocomplete-highlighter="customHighlighter" class="o-autocomplete"  id="my-autocomplete">
+            <div data-o-component="o-autocomplete" data-o-autocomplete-source="customSuggestions" data-o-autocomplete-highlighter="customHighlighter" class="o-autocomplete" >
                <!-- If the JavaScript executes, then this input will be progressively enhanced to an autocomplete component -->
-                <input required type="text" name="text-example">
+                <input required type="text" name="text-example" id="my-autocomplete">
            </div>
```

notlee replied:
Would be a good thing to shout about in the readme too :)

JakeChampion replied:
Done in [`eb54004` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/eb54004c313d90eeb5b012e8bb8eb61f422d6857)

notlee commented on src/js/autocomplete.js:
```diff
@@ -6,9 +13,187 @@ class Autocomplete {
+		let timeout = null;
+		clearButton.addEventListener('click', function clearInput() {
+			input.value = '';
+			clearButton.parentElement.removeChild(clearButton);
```
The clear button should also remove the loading state, in-case of a slow response.

JakeChampion replied:
Good idea -- I've done this in [`bb505f9` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/bb505f9661c87531115f1ab776811e12eb1467ce)

notlee commented on src/js/autocomplete.js:
```diff
@@ -6,9 +13,187 @@ class Autocomplete {
+		// <div class="o-autocomplete__menu-loading-container">
+		//     <div class="o-autocomplete__menu-loading"></div>
+		// </div>
+		const loadingContainer = document.createElement('div');
```
Non-blocking suggestion: breaking the constructor up with private functions would make it a bit easier for future us to read and work on it

JakeChampion replied:
Done in [`27c5216` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/27c521694db2f19d787b9c4a9e0c8ff747a49ccc)

notlee commented on src/js/autocomplete.js:
```diff
@@ -6,9 +13,187 @@ class Autocomplete {
+			}
+		}
+		function hideLoadingPane() {
+			//TODO: display none it
```
I think this todo has been todone

JakeChampion replied:
```suggestion
```

notlee commented on src/js/autocomplete.js:
```diff
@@ -6,9 +13,187 @@ class Autocomplete {
+		}
+
+		// o-autocomplete has a UI design option to highlight characters in the suggestions.
+		// The product using o-autocomplete decides upon the logic behind which characters to highlight and which to not.
```
Do we expect highlighter rules will often change? If not, I think we might want to provide a default highlighter following the rules in the current design

JakeChampion replied:
I reckon not, I reckon we change this to have a specific highlighter which can not be overridden. If a product wants a different highlight method, that would then need to get design approval first 👍

JakeChampion replied:
Removed this customisation feature in [`8259cee` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/8259cee79c501892cd5f53ca1779109dd9a82fa8)

notlee reviewed (commented):
This is an excellent component, nice work! :D I haven't reviewed tests yet as there are some to-dos in there

JakeChampion committed:
capitalise Origami in readme

Co-authored-by: Lee Moody <notlee@users.noreply.github.com>

JakeChampion committed:
remove passive voice

Co-authored-by: Lee Moody <notlee@users.noreply.github.com>

JakeChampion committed:
update sass import example to import the `main.scss` file

JakeChampion committed:
convert mustache comment to html comment to show it in the built demo html

JakeChampion committed:
remove todo comment which has been done

notlee committed:
Remove the `o-autocomplete__container` element.

To reduce code complexity and css specificity.

The child `o-autocomplete__wrapper` element can be used instead.
Since it is created by the alphagov dependency there's an increased
risk of conflicts but we are already styling elements created by the
dependency, we can mitigate this by pinning the dependency version
number.

notlee committed:
Position the menu-loading-container absolutely.

So it does not push page content down, and takes up the full width
now the container is removed (see previous commit). Also applies a
background colour so page content behind the loading container is
hidden.

notlee committed:
Remove Sass tests

Since there is only one primary mixin with no options -- Percy can
catch visual changes

notlee committed:
Fix bug which caused an error on init when no source option is given

Introduced: a6d3a5fe774496ffc75fe82f20c662d0b302d5bc

JakeChampion committed:
Remove the ability to override how the suggestions should be highlighted as the highlighting is a design decision and a new highlighting design should get approval before being used in a product

JakeChampion committed:
extract out from the constructor the clear button and loading pane logic and place them into private methods

JakeChampion committed:
When pressing the clear button, hide the loading pane it is showing

JakeChampion committed:
document the requirement of an `id` attribute on the `input`

JakeChampion committed:
removed references to the origami initialisation page

JakeChampion committed:
improve the dynamic suggestions example to show how to configure o-autocomplete with the suggestion function

JakeChampion committed:
improve the jsdoc for autocomplete options

JakeChampion committed:
add jsdoc to the example in the readme to help describe the types being used

JakeChampion committed:
add tests for the dynamic autocomplete suggestions option

JakeChampion committed:
Add example to readme on how to use with o-forms

JakeChampion commented:
@chee & @notlee - I've made changes based on the feedback, this is now ready for another review 👍

JakeChampion committed:
update the automated tagging system to be able to trigger the publishing workflows

notlee commented on README.md:
```diff
@@ -16,25 +23,67 @@ Check out [how to include Origami components in your project](https://origami.ft
+    <span class="o-forms-title">
+        <span class="o-forms-title__main">Select your country</span>
+    </span>
+    <span class="o-forms-input o-forms-input--select o-forms-input--text">
```
Think there is an excess modifier here (haven't tested that)
```suggestion
    <span class="o-forms-input o-forms-input--select">
```

notlee commented on README.md:
```diff
@@ -16,25 +23,67 @@ Check out [how to include Origami components in your project](https://origami.ft
+
+### Use with o-forms
+
+To have styling for labels, you will need to use o-forms as part of the autocomplete implementation.
```
🙌 
```suggestion
To have styling for labels, you will need to use [o-forms](https://registry.origami.ft.com/components/o-forms) as part of the autocomplete implementation.
```

notlee commented on src/js/autocomplete.js:
```diff
@@ -1,14 +1,247 @@
+		}
+		return [character, shoudHighlight];
+	});
+}
```
```suggestion
	return result.map(function(character, index) {
		let shouldHighlight = true;
		const hasMatched = matchIndex > -1;
		const characterIsWithinMatch = index >= matchIndex && index <= matchIndex + query.length - 1;
		if (hasMatched && characterIsWithinMatch) {
			shouldHighlight = false;
		}
		return [character, shouldHighlight];
	});
}
```

notlee commented on src/js/autocomplete.js:
```diff
@@ -1,14 +1,247 @@
+	});
+}
+
+/**
+ * Create the DOM tree which corresponds to
+ * <div class="o-autocomplete__menu-loading-container">
+ * 	<div class="o-autocomplete__menu-loading"></div>
+ * </div>
+ * @returns {HTMLDivElement} The loading container as a HTMLDivElement
+ */
+function createLoadingContainer() {
+	const loadingContainer = document.createElement('div');
+	loadingContainer.classList.add('o-autocomplete__menu-loading-container');
+	const loading = document.createElement('div');
+	loading.classList.add('o-autocomplete__menu-loading');
+	loadingContainer.appendChild(loading);
+	return loadingContainer;
+}
```
Non-blocking suggestion. I was curious if we could improve this function to remove the need for the DOM tree comment and think we can 😄 Could maybe do a similar thing for `createClearButton` too

```suggestion
/**
 * Create DOM for the loading container.
 * @returns {HTMLDivElement} The loading container.
 */
function createLoadingContainer() {
	return document.createRange().createContextualFragment(`
		<div class="o-autocomplete__menu-loading-container">
			<div class="o-autocomplete__menu-loading"></div>
		</div>
	`).querySelector(':first-child'); // IE does not support firstElementChild
}
```

JakeChampion replied:
Nice!

notlee commented on src/js/autocomplete.js:
```diff
@@ -1,14 +1,247 @@
+				customSuggestions(query, callback);
+			};
+			const input = autocompleteEl.querySelector('input');
+			const id = input.getAttribute('id');
```
Question: is this id required for accessible-autocomplete to work? If so should we throw our own error here?

JakeChampion replied:
It is required, accessible-autocomplete throws an error if it is missing. We can throw our own error instead, yeah 👍

JakeChampion replied:
Done in [`da94664` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/da9466432c1b585cde53506a6488f8fdd228bfea)

notlee commented on src/js/autocomplete.js:
```diff
@@ -1,14 +1,247 @@
+				selectElement: element,
+				defaultValue: '',
+			}, this.options);
+			options.autoselect = false; // TODO: Find out if we should allow autoselect/hinting in the input
```
I forget what these features are. Should we? :D

notlee replied:
Reading this comment back it sounds enthusiastic for adding features when actually let's not if we don't need them yet 😅

JakeChampion replied:
```suggestion
			options.autoselect = false;
```

JakeChampion commented on demos/src/static/static.mustache:
```diff
@@ -0,0 +1,268 @@
+    <span class="o-forms-title">
+        <span class="o-forms-title__main">Select your country</span>
+    </span>
+    <span class="o-forms-input o-forms-input--select o-forms-input--text">
```
```suggestion
    <span class="o-forms-input o-forms-input--select">
```

JakeChampion committed:
remove input text css modifier on the select elements

Co-authored-by: Lee Moody <notlee@users.noreply.github.com>

notlee commented on src/js/autocomplete.js:
```diff
@@ -1,14 +1,247 @@
+
+	/**
+	 * Show the loading panel
+	 * @private
```
Non-blocking suggestion/question: I was about to comment that these should be private, but see you handled that with JSDoc :) I worry a bit that projects may come to rely on them accidentality if users don't consult the docs/have hinting. My preference would be to additionally indicate they are private by prepending an underscore to their name, as some other components do, or take them off the prototype and pass in the autocomplete instance to make them actually private

I think it's fine as is though, I wouldn't feel toooo guilty telling any projects that break due to relying on methods documented as private to not do that, what are your thoughts?

JakeChampion replied:
Happy to have them pulled out of the prototype and become truly private functions which take the instance as an argument. Talking about private methods -- Private class methods are going to be added to the ecma-2022 specification, they are already at stage 4 of the process -- https://github.com/tc39/proposals/blob/master/finished-proposals.md#finished-proposals

JakeChampion replied:
Done in [`a97cf23` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/a97cf2316cb1e52ae17314c09641196918ac7187)

JakeChampion commented on README.md:
```diff
@@ -1,11 +1,18 @@
 - [Usage](#usage)
 - [Markup](#markup)
+    - [For a static set of suggestions](#for-a-static-set-of-suggestions)
+    - [For a dynamic set of suggestions](#for-a-dynamic-set-of-suggestions)
```
```suggestion
    - [For a dynamic set of suggestions](#for-a-dynamic-set-of-suggestions)
    - [Use with o-forms](#use-with-o-forms)
```

JakeChampion committed:
update table of contents

notlee reviewed (dismissed):
Thanks for making those changes, I think this is almost ready to go! 🎉 I've spotted a couple of issues with a second pass including a quick IE11 test. Looks like there are some JS errors and the default clear button isn't hidden:

![Screenshot 2021-06-18 at 13 28 57](https://user-images.githubusercontent.com/10405691/122560967-29ec5700-d039-11eb-933b-cc832d7151fc.png)

Most the other comments are non-blocking suggestions / questions for my own benefit :)

JakeChampion committed:
add link to o-forms

Co-authored-by: Lee Moody <notlee@users.noreply.github.com>

JakeChampion committed:
refactor createLoadingContainer

Co-authored-by: Lee Moody <notlee@users.noreply.github.com>

JakeChampion committed:
remove TODO comment which we are not planning to do

JakeChampion committed:
fix typo in variable name

Co-authored-by: Lee Moody <notlee@users.noreply.github.com>

JakeChampion committed:
make private methods actually private rather than relying on documentation

JakeChampion committed:
Add custom error message if the id attribute is not set

JakeChampion committed:
add npm test script which runs obt t

JakeChampion committed:
use stable remark lint config

JakeChampion committed:
use obt 11

JakeChampion committed:
hide IE 11's native clear button

JakeChampion committed:
mark String.prototype.startsWith as required

JakeChampion committed:
make loading container work in ie11

:first-child does not match elements with no parents such as the first element in a document fragment

JakeChampion committed:
empty commit to trigger ci

JakeChampion committed:
refactor createClearButton to use a documentfragment instead of constructing it by hand

notlee commented:
Spotted some more issues!

At some point we lost the padding which prevented input overflowing the button. Is that what we were using the container element for? 😳 
![Screenshot 2021-06-18 at 17 01 30](https://user-images.githubusercontent.com/10405691/122589068-ba399480-d057-11eb-9188-10447d97d4df.png)

Weird, why does "country" and "one more item" have no space. Is "one more item" the clear button, and is that how it should be announced? I'm all questions. No answers.
![Screenshot 2021-06-18 at 17 00 26](https://user-images.githubusercontent.com/10405691/122589356-0edd0f80-d058-11eb-9054-ac370ba20414.png)

The results are read as letters with spaces, not words 😬 Because of the span elements used to highlight. Is that markup generated by us or `accessible-autocomplete`? I'd guess the answer is `aria-label`.
![Screenshot 2021-06-18 at 17 01 17](https://user-images.githubusercontent.com/10405691/122589504-3cc25400-d058-11eb-8c48-2b1c867012d5.png)
![Screenshot 2021-06-18 at 17 12 11](https://user-images.githubusercontent.com/10405691/122589576-56639b80-d058-11eb-8805-c4305a053039.png)

notlee commented:
>At some point we lost the padding which prevented input overflowing the button. Is that what we were using the container element for? 😳

Do we have a Percy test for an autocomplete with input / options dropdown. If not, that'd be great to catch any regressions with the suggestions UI in the future!

JakeChampion committed:
Use our fork of accessible-autocomplete which works for async suggestions when the query is deleted

notlee commented:
>Weird, why does "country" and "one more item" have no space.

Banana completed
https://twitter.com/rowanmanning/status/1069549765828972544?s=20

JakeChampion committed:
make the highlighted text be read out correctly by assistive tech

JakeChampion commented:
> At some point we lost the padding which prevented input overflowing the button. Is that what we were using the container element for? 😳

Not sure, I'll need to look into this.



> Weird, why does "country" and "one more item" have no space. Is "one more item" the clear button, and is that how it should be announced? I'm all questions. No answers.

I'm not sure where "one more item" comes from but can confirm I hear it to. The clear button has a title attribute `Clear input` which should be read out


>The results are read as letters with spaces, not words 😬 Because of the span elements used to highlight. Is that markup generated by us or accessible-autocomplete? I'd guess the answer is aria-label.

Good spot, that is fixed in [`afb25c4` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/afb25c4aef37646b166b93e4c7f09582cf43a104)

JakeChampion committed:
Ensure that .o-autocomplete__input's padding is used instead of o-forms

JakeChampion commented:
> > At some point we lost the padding which prevented input overflowing the button. Is that what we were using the container element for? 😳
> 
> Not sure, I'll need to look into this.
> 

This looks to be a battle between o-forms and o-autocomplete, fixed in [`bc19cbe` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/bc19cbe5a584837e14a103a3f83c4f63680fce18)

notlee committed:
Restore container element.

We decided to initalise accessible-autocomplete with the root
o-autocomplete element instead of a child "container"
element to simplify the codebase a little and remove some
CSS specificity.

See: 7f90182979a4a184123770bb9d2da13915936cbf

However, this broke the "static" demo which shows the variant that
relies on a `select` input for its options. This is because
accessible-autocomplete generates different DOM for both variants
(the wrapper element is no longer a direct child of o-autocomplete).

So this commit restores the container element but also ensures
any generated o-autocomplete DOM (i.e. the clear button) is placed
outside the container. Now o-autocomplete's layout is independent
of accessible-autocomplete's markup.

notlee committed:
Remove console.log

notlee committed:
Do not clear suggestions on input click.

Upgrades `@financial-times/accessible-autocomplete`, which contains
the bug fix.

notlee committed:
Restore static demo to include `text` and `select` form modifiers.

The core experience is a `select` input which is enhanced using
a `text` input. Both o-forms modifiers are needed to handle each
case. This isn't ideal -- o-forms wasn't intended to be used this
way and there may be conflicting styles in the future.

notlee committed:
Improve source error handling / tests

- Error if no source is given (either a select element or function)
- Fix test for an input missing an id

notlee committed:
convert o-forms markup comments from html to mustache

they are for authors not users

notlee committed:
Remove todo not on o-form demo markup.

It may be an issue if `o-forms` decides to style an `input` with
`.o-forms-input--select` or a `select` with `.o-forms-input--text`.
We don't think this is likely to be a problem and may be caught
with visual regression testing.

JakeChampion commented on origami.json:
```diff
@@ -9,19 +8,41 @@
 		"documentClasses": "",
-		"dependencies": []
+		"dependencies": [
+			"o-forms@9.0.0-beta.1",
+			"o-normalise@3.0.0-beta.0",
+			"o-utils@2.0.0-beta.0"
```
```suggestion
			"o-forms@^9.0.0",
			"o-normalise@^3.0.0",
			"o-utils@^2.0.0"
```

JakeChampion replied:
Done in [`49b0363` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/49b0363ed2605eeb4fbcb4159256fdd546ab0680)

JakeChampion commented on package.json:
```diff
@@ -1,39 +1,57 @@
-	"browser": "main.js",
 	"peerDependencies": {
-		"@financial-times/o-brand": "^4.0.0-beta.0"
+		"@financial-times/o-brand": "prerelease",
+		"@financial-times/o-colors": "prerelease",
+		"@financial-times/o-forms": "prerelease",
+		"@financial-times/o-grid": "prerelease",
+		"@financial-times/o-icons": "prerelease",
+		"@financial-times/o-loading": "prerelease",
+		"@financial-times/o-normalise": "prerelease",
+		"@financial-times/o-spacing": "prerelease",
+		"@financial-times/o-typography": "prerelease",
+		"@financial-times/o-visual-effects": "prerelease"
```
Need to update to latest stable versions

JakeChampion replied:
Done in [`18d324b` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/18d324b50e01dbd59236f28b1b31e1b29bb3bb61)

JakeChampion commented on src/js/autocomplete.js:
```diff
@@ -1,14 +1,264 @@
+// We use our own fork of accessible-autocomplete because of this bug -- https://github.com/alphagov/accessible-autocomplete/issues/210#issuecomment-845860595
+// Our fork uses this change -- https://github.com/alphagov/accessible-autocomplete/pull/491
+// If the bug is fixed in accessible-autocomplete then we can stop using our fork
```
We should also include this bug fix -- https://github.com/alphagov/accessible-autocomplete/pull/496

JakeChampion replied:
This was done in [`0f25101` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/0f2510189a8853e789e4616fbae7b7d4797e479d)

JakeChampion committed:
update to latest version of @financial-times/accessible-autocomplete

JakeChampion committed:
update the comment about using a fork of accessible-autocomplete

JakeChampion committed:
update to the latest major version of @testing-library/dom

We are not using any of the features which were removed or changed in the new version.

The new version also removes the need for the hack.js file we needed to use this is because, the package is using the latest major versio of pretty-format, which contains this update https://github.com/facebook/jest/pull/11569

JakeChampion committed:
update origami type to module from component

JakeChampion committed:
use latest stable dependencies for demos

JakeChampion committed:
remove unused o-forms dependency

JakeChampion committed:
update to latest stable versions of origami dependencies

JakeChampion commented:
> > At some point we lost the padding which prevented input overflowing the button. Is that what we were using the container element for? 😳
> 
> Do we have a Percy test for an autocomplete with input / options dropdown. If not, that'd be great to catch any regressions with the suggestions UI in the future!

I don't think that is possible with our current Percy setup unfortunately because we are using ["percy snapshot"](https://docs.percy.io/docs/command-line-client#percy-snapshot-snapshotdirectory) within [origami-percy](https://github.com/Financial-Times/origami-percy/blob/f3a3b4af994efc6dcad4d21aefbc2de48d3ef15a/src/index.js#L67) and I think that snapshots as soon as possible. I believe we would need to move from the generic origami-percy action and use a custom percy script if we want to test more interactions with a component, something like [percy's puppeteer](https://docs.percy.io/docs/puppeteer) api

JakeChampion commented on origami.json:
```diff
@@ -1,6 +1,5 @@
 {
-	"origamiType": "component",
-	"origamiCategory": "components",
+	"origamiType": "module",
```
```suggestion
	"origamiType": "component",
```

JakeChampion committed:
update origami.json to component

JakeChampion commented:
@notlee I added a release:major tag, hopefully that will release v1 🤞

notlee reviewed (dismissed):
So close! but IE11 seems unhappy again

![Screenshot 2021-07-02 at 17 44 08](https://user-images.githubusercontent.com/10405691/124305540-25f22600-db5d-11eb-9f4d-fa3991ea5930.png)

JakeChampion committed:
add missing required browser features to get component to work in ie 11

JakeChampion commented:
@notlee it now works in IE11 once again 👍 
![image](https://user-images.githubusercontent.com/1569131/124562795-57a60e00-de37-11eb-870b-1429163e7b7d.png)
